postpic/helper.py

```diff
@@ -192,6 +198,114 @@ def linear2polar(x, y):
+
+    `scipy.ndimage.map_coordinates` is slow for large datasets. Speed improvement can be
+    achieved by
+
+     * Splitting the data into chunks
+     * Performing the transformation of chunks in parallel
+
+    New parameters:
+
+    chunklen: Size of the chunks in pixels per axis. Default: None
+        Special values:
+            None: Automatic (Chooses a default based on number of dimensions)
+            0: Do not split data into chunks. (implicitly sets threads==1)
+
+    threads: Number of threads. Default: None
+        None: Automatic (One thread per available processing unit)
+
+    """
+
+    # this part is taken without change from scipy's serial implementation
+    if order < 0 or order > 5:
+        raise RuntimeError('spline order not supported')
+    input = np.asarray(input)
+    if np.iscomplexobj(input):
+        raise TypeError('Complex type not supported')
+    coordinates = np.asarray(coordinates)
+    if np.iscomplexobj(coordinates):
+        raise TypeError('Complex type not supported')
+    output_shape = coordinates.shape[1:]
+    if input.ndim < 1 or len(output_shape) < 1:
+        raise RuntimeError('input and output rank must be > 0')
+    if coordinates.shape[0] != input.ndim:
+        raise RuntimeError('invalid shape for coordinate array')
+    mode = _ni_support._extend_mode_to_code(mode)
+    if prefilter and order > 1:
+        filtered = spline_filter(input, order, output=np.float64)
+    else:
+        filtered = input
+
+    # return value of `_ni_support._get_output` changed between scipy versions, code here is
+    # adapted to work with both
+    output = _ni_support._get_output(output, input, shape=output_shape)
+    retval = output
+    if isinstance(output, tuple):
+        output, retval = output
+
+    # below here there is the new code for splitting into chunks and parallel execution
+    if chunklen is None:
+        # set defaults
+        chunklen = 128
+        if output.ndim < 3:
+            chunklen = 1024
+
+    def chunk_arguments(filtered, coordinates, output):
+        chunks = []
+        for axis in range(output.ndim):
+            chunkstarts = np.arange(0, output.shape[axis], chunklen)
+            chunkends = chunkstarts + chunklen
+            chunkends[-1] = output.shape[axis]
+            chunks.append([slice(start, stop) for start, stop in zip(chunkstarts, chunkends)])
+
+        for chunk in itertools.product(*chunks):
+            sub_coordinates = coordinates[(slice(None),) + chunk].copy()
+            filtered_region = []
+            for in_axis in range(filtered.ndim):
+                c = sub_coordinates[in_axis, ...]
+                cmin = max(0, int(np.floor(np.min(c)))-5)
+                cmax = min(filtered.shape[in_axis], int(np.ceil(np.max(c)))+5)
+                sub_coordinates[in_axis, ...] -= cmin
+                filtered_region.append(slice(cmin, cmax))
+            sub_filtered = filtered[tuple(filtered_region)]
+            sub_output = output[chunk]
```

Comment:
Die `sub-outputs` ueberlappen bei ja ein bisschen (`-5`, `+5`), sind aber views auf dasselbe numpy array. Die verschiedenen Threads schreiben also nachher in verschiedene views desselben numpy arrays. Das finde ich ueberaus elegant, weiss aber nicht wie das mit dem GIL funktioniert. Wir der GIL wirklich komplett released oder haengt der an dem base-array? Insbesondere da die `sub_output`s ja ueberlappen ist mir unklar wie das funktioniert und ob dort nicht etwas speed verschenkt wird. Oder habe ich da etwas falsch verstanden?

Reply:
Ich denke, dass es lohnen wuerde ein beispiel zu schrieben (unter `examples/`), welches als teil der tests laeuft und die Performance der Parallelen im vergleich zu Sequenziellen version ausgibt. Aehnlich wie das schon bei den Partilce Properties gemacht ist.

Reply:
Es sind nicht die Outputs die überlappen, sondern mit den -5/+5 wird der Input-Bereich für jeden einzelnen Thread etwas ausgedehnt um Randeffekte auszuschließen. Die Outputs sind disjunkt. Die zu grunde liegende C-Routine gibt das GIL frei, so dass der code wirklich parallel läuft.

Ich schau mal, dass ich da bei Gelegenheit noch ein Beispiel mit Benchmark baue und einen Test schreibe, der Seriell vs Parallel vergleicht.